### ISSUE TITLE
Cache directory preserves original casing of repo_id → duplicate caches on case-sensitive filesystems (Linux)

### ISSUE THREAD

**Issue by julio75012** [huggingface/huggingface_hub#3838]
### Describe the bug

`from_pretrained` and `hf_hub_download` build the local cache directory using the **exact casing** of the provided `repo_id`.

On case-sensitive filesystems (Linux), different capitalizations of the same model create separate cache folders and trigger duplicate downloads. The Hugging Face Hub treats repository IDs as case-insensitive.

### Steps to reproduce

```python
from transformers import AutoModel, AutoTokenizer

model_mixed = "deepseek-ai/DeepSeek-OCR"
model_lower = "deepseek-ai/deepseek-ocr"   # same model

# On Linux these create two different folders:
# ~/.cache/huggingface/hub/models--deepseek-ai--DeepSeek-OCR
# ~/.cache/huggingface/hub/models--deepseek-ai--deepseek-ocr

AutoModel.from_pretrained(model_mixed, trust_remote_code=True)
# Running with model_lower downloads again
```

### Expected behavior
Normalize the `repo_id` (e.g. lowercase) when generating the cache folder name so all case variations resolve to the **same** cache location.

### Actual behavior
Cache path preserves original casing → duplicate folders on Linux.

### Environment
- OS: Linux (case-sensitive filesystem)
- `huggingface_hub` version: `python -c "import huggingface_hub; print(huggingface_hub.__version__)"`
- `transformers` version: `python -c "import transformers; print(transformers.__version__)"`

### Suggested fix
Lowercase the `repo_id` for cache-key generation inside `huggingface_hub` (while keeping original casing for Hub API calls).

This eliminates wasted disk space from common casing variations in model names.

**Comment by julio75012** [comment]
fix proposed here: https://github.com/huggingface/huggingface_hub/pull/3839

**Comment by Wauplin** [comment]
Hi @julio75012 , this is a known topic that we don't really want to solve. If user always passes the same repo id, they will always use the same cache folder. The duplicate cache issue happens only when user explicitly hardcodes 2 different casing for the same repo, which we consider to be a corner case. We also don't want to use the resolved repo id to build the cache folder as it would invalidate everyone's cache in case of repo author renames their repo. I will therefore close this issue and PR without merging. 

Note that the PR you've suggested would invalidate the existing cache of everyone using `huggingface_hub`. Some users have TB+ of local cache and would be quite annoyed by that ^^

**Comment by hoedt** [comment]
@Wauplin I respectfully disagree with your assessment. In an interactive session I have often been surprised to see download bars appear because I did not correctly capitalise the name of the model (probably because I sometimes do not hit the shift key properly). Also, this ignores use cases like sharing a huggingface cache on a multi-user system (where it is typically very hard to impose a global naming rule for everyone) or relying on different tools/repos in a single project (where each tool uses the hub with their own naming convention).

I think the problem of bloating the cache with duplicate information (and the time lost due to repeated downloading) can and should be taken seriously. I agree that the proposed PR is not a good solution, but I would argue that it should not be too hard to find a simple workaround for this issue, without invalidating TB of (likely duplicate) data in the cache of some users.

**Comment by julio75012** [comment]
I again wasted time downloading again something that I already had on my disk

**Comment by julio75012** [comment]
@Wauplin could you please clarify who hardcoded 

model_name = "deepseek-ai/deepseek-ocr"
and 
model_name = "deepseek-ai/DeepSeek-OCR"

here is a extract of my cached directories:
```
drwxrwxr-x 6 jules jules 4096 Apr 20 15:10 models--deepseek-ai--deepseek-ocr
drwxrwxr-x 6 jules jules 4096 Mar 24 17:20 models--deepseek-ai--DeepSeek-OCR
```

just for information, I did not hardcode anything, I just download the models. therefore there is probably some duplication of data (for case sensitivity reasons...) on some servers

**Comment by Wauplin** [comment]
> just for information, I did not hardcode anything, I just download the models. 

How did you downloaded them? Which command lines or scripts did you run? 

For context, here was my response a few days ago for a similar report: https://github.com/huggingface/huggingface_hub/issues/4105#issuecomment-4250713696.

**Comment by julio75012** [comment]
script 1:
```
import sys
import transformers.models.llama.modeling_llama as modeling_llama
from transformers import AutoModel, AutoTokenizer
import torch
import os

model_name = "deepseek-ai/DeepSeek-OCR"

tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
model = AutoModel.from_pretrained(
    model_name,
    _attn_implementation="eager", 
    trust_remote_code=True,
    use_safetensors=True,
)
```

script 2 (same but lowercase for model_name):
```
import sys
import transformers.models.llama.modeling_llama as modeling_llama
from transformers import AutoModel, AutoTokenizer
import torch
import os

model_name = "deepseek-ai/deepseek-ocr"

tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
model = AutoModel.from_pretrained(
    model_name,
    _attn_implementation="eager", 
    trust_remote_code=True,
    use_safetensors=True,
)
```

I would like to know who "hardcodes 2 different casing for the same repo" (quoting you): for sure, not me.
And, if on a server on which it is uploaded, if there is duplicated data (one in deepseek-ocr and another one in DeepSeek-OCR).

**Comment by Wauplin** [comment]
I think there is a misunderstanding here. What I meant is that if different model names are hardcoded differently in 2 separate scripts, then they will be downloaded to 2 different locations on disk even if data is strictly the same. It's suboptimal but won't be fixed anytime soon (cf https://github.com/huggingface/huggingface_hub/issues/4105#issuecomment-4250713696). The correct solution is to define `model_name = "deepseek-ai/DeepSeek-OCR"` in both scripts.
